### PR TITLE
build.d: Add ability to determine the host C++ compiler on Windows

### PR DESCRIPTION
**Pull request by JinShil**

### CODE REVIEW

File: src/build.d [hunk 1]
```diff
@@ -820,6 +821,29 @@ auto detectModel()
...
+{
+    version(Posix)
+        return "c++";
+    else version(Windows)
+    {
+        immutable model = detectModel;
+        if (model == "32")
+            return "dmc";
+        else if (model == "64")
+            return `vcbuild\msvc-dmc`;
```
Comment (JinShil) [inline src/build.d]: See https://github.com/dlang/dmd/blob/fd66f64bf8f8720b5248a8c3f836c899d55e4676/src/win64.mak#L24

`msvc-dmc` can be found in https://github.com/dlang/dmd/tree/master/src/vcbuild

**Review by PetarKirov — APPROVED** [review]

File: src/build.d (hunk 1 above)
Comment (JinShil) [reply src/build.d]: My understanding is that for 32-bit code, the makefiles us DMC:  https://github.com/dlang/dmd/blob/fd66f64bf8f8720b5248a8c3f836c899d55e4676/src/win32.mak#L88

Comment (wilzbach) [reply src/build.d]: I think you need to make sure that this file is compiled.

```d
Dependency dependency = {
        target: env["G"].buildPath("msvc-dmc").exeName,
        sources: ["vcbuild\msvc-dmc"],
};
dependency.run();
```

(though if you put this in `$G`, make sure `mkdirRecurse(g);` (also part of `parseEnvironment` is called first)).

Comment (JinShil) [reply src/build.d]: If you're implying that we should be using MSVC for both 32 and 64-bit code, well, I'm trying not to rock the boat.

Comment (JinShil) [reply src/build.d]: > I think you need to make sure that this file is compiled.

Yes, I'm just working through the script step by step.  I'll get there.

Comment (wilzbach) [reply src/build.d]: And looking a bit more at the win64.mak file, it seems like it also requires to build the `vcbuild\msvc-lib` file

File: src/build.d [hunk 2]
```diff
@@ -236,6 +236,19 @@ auto opTabGen()
...
+version(Windows)
+{
+    // Build the msvc-dmc compiler wrapper
+    void buildMsvcDmc()
+    {
+        Dependency dependency = {
+            target: env["G"].buildPath("msvc-dmc").exeName,
+            sources: [`vcbuild\msvc-dmc`],
+        };
+        dependency.run();
```
Comment (wilzbach) [inline src/build.d]: `return dependency.target`

File: src/build.d [hunk 3]
```diff
@@ -836,7 +849,10 @@ auto getHostCXX()
         if (model == "32")
             return "dmc";
         else if (model == "64")
+        {
+            buildMsvcDmc();
             return `vcbuild\msvc-dmc`;
```
Comment (wilzbach) [inline src/build.d]: `return buildMsvcDmc()` (otherwise you won't use the built msvc-dmc.exe)

File: src/build.d (hunk 1 above)
Comment (JinShil) [reply src/build.d]: Added `msvc-lib`, but they're not building yet, that's a bigger task than I intended to tackle today.

File: src/build.d (hunk 3 above)
Comment (JinShil) [reply src/build.d]: I added `msvc-dmd` and `msvc-lib` dependencies for `cxxBackend`.  I think I need `processEnvironment` to run first before those can be built.

File: src/build.d [hunk 4]
```diff
@@ -270,6 +293,15 @@ auto dBackend()
 auto cxxBackend()
 {
     Dependency[] dependencys;
```
Comment (wilzbach) [inline src/build.d]: `dependencies` - sorry for the typo :O

File: src/build.d (hunk 1 above)
Comment (wilzbach) [reply src/build.d]: Okay, but I think you still need sth. like `buildMsvcDmc().target` here to return the correct path.

File: src/build.d (hunk 4 above)
Comment (JinShil) [reply src/build.d]: I will fix that later.  There are a few places it needs to change and I'd like to keep the diff low.

**Review by wilzbach — APPROVED** [review]

**Pull request merged**